pcap-linux.c

```diff
@@ -1469,42 +1474,60 @@ pcap_activate_linux(pcap_t *handle)
 		 * Success.
 		 * Try to use memory-mapped access.
 		 */
-		switch (activate_mmap(handle, &status)) {
+		ret = activate_mmap(handle, &status);
+		if (ret == -1) {
+			/*
+			 * We failed to set up to use it, or the
+			 * kernel supports it, but we failed to
+			 * enable it.  status has been set to the
+			 * error status to return and, if it's
+			 * PCAP_ERROR, handle->errbuf contains
+			 * the error message.
+			 */
+			goto fail;
+		}
 
-		case 1:
+
+		if (ret == 1) {
 			/*
 			 * We succeeded.  status has been
 			 * set to the status to return,
 			 * which might be 0, or might be
 			 * a PCAP_WARNING_ value.
-			 *
-			 * Set the timeout to use in poll() before
-			 * returning.
 			 */
-			set_poll_timeout(handlep);
-			return status;
-
-		case 0:
 			/*
-			 * Kernel doesn't support it - just continue
-			 * with non-memory-mapped access.
+			 * Now that we have activated the mmap ring, we can
+			 * set the correct protocol.
 			 */
-			break;
-
-		case -1:
+			if ((status2 = iface_bind(handle->fd, handlep->ifindex,
+			     handle->errbuf, pcap_protocol(handle))) != 0) {
+				status = status2;
+				goto fail;
+			}
 			/*
-			 * We failed to set up to use it, or the
-			 * kernel supports it, but we failed to
-			 * enable it.  status has been set to the
-			 * error status to return and, if it's
-			 * PCAP_ERROR, handle->errbuf contains
-			 * the error message.
+			 * Set the timeout to use in poll() before
+			 * returning.
 			 */
-			goto fail;
+			set_poll_timeout(handlep);
+			return status;
 		}
+
+		/*
+		 * Kernel doesn't support it (mmap) - just continue
+		 * with non-memory-mapped access.
+		 */
 #endif /* HAVE_PACKET_RING */
 	}
 
+	/*
+	 * We need to set the correct protocol.
+	 */
+	if ((status2 = iface_bind(handle->fd, handlep->ifindex,
```

Comment:
If the "any" device is opened, is `handlep->ifindex` 0?  And does that result in the socket being bound to a protocol but not to an interface (as I suspect it will, from a quick look at af_packet.c), so that packets from all interfaces show up (e.g., your main interface and the loopback interface)?

Reply:
My understanding is that ifindex would be -1, and I map that back to 0 in iface_bind.  The man page for "packet" says about ifindex "0 matches any interface"

Reply:
Seems to work in my tests.